### PR TITLE
Check for empty message in Envelope class

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
I've also removed all getters in the properties

### CODE REVIEW

File: src/core/Akka/Actor/ActorCell.cs [hunk 1]
```diff
@@ -308,33 +308,31 @@ public virtual void SendMessage(IActorRef sender, object message)
...
-
+                    Serializer serializer = _systemImpl.Serialization.FindSerializerFor(message.Message);
+                    byte[] serialized = serializer.ToBinary(message.Message);
 
                     var manifestSerializer = serializer as SerializerWithStringManifest;
                     if (manifestSerializer != null)
                     {
                         var manifest = manifestSerializer.Manifest(serialized);
-                        message = _systemImpl.Serialization.Deserialize(serialized, manifestSerializer.Identifier, manifest);
+                        message = new Envelope(_systemImpl.Serialization.Deserialize(serialized, manifestSerializer.Identifier, manifest), message.Sender);
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/ActorCell.cs]: Scala version has a bit different implementation
https://github.com/akka/akka/blob/4acc1cca6a27be0ff80f801de3640f91343dce94/akka-actor/src/main/scala/akka/actor/dungeon/Dispatch.scala#L138

I should check for null, and then throw an exception

File: src/core/Akka.FSharp.Tests/ApiTests.fs [hunk 2]
```diff
@@ -121,8 +121,8 @@ let ``actor that accepts _ will receive string message`` () =
     response
     |> equals "SomethingToReturn"
 
-[<Fact>]
-// SUCCEEDS
+//[<Fact>]
+// FAILS
```
Comment (alexvaluyskiy) [inline src/core/Akka.FSharp.Tests/ApiTests.fs]: I've discussed about this test with @Horusiath . He told me, that I should remove this test if it is not actual at the moment

File: src/core/Akka/Actor/Message.cs [hunk 3]
```diff
@@ -5,26 +5,43 @@
...
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+
+            Message = message;
+            Sender = sender != ActorRefs.NoSender ? sender : system.DeadLetters;
+        }
+
+        public Envelope(object message, IActorRef sender)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/Message.cs]: I should have made this constructor as internal, but I did not manage to do it. I've got some conflicts in Persistence

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt [hunk 4]
```diff
@@ -101,6 +101,7 @@ namespace Akka.Actor
         public void ReserveChild(string name) { }
         public void Restart(System.Exception cause) { }
         public void Resume(System.Exception causedByFailure) { }
+        public virtual void SendMessage(Akka.Actor.Envelope message) { }
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: Do we need this? Why not have the other `SendMessage` method construct the envelope?

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt [hunk 5]
```diff
@@ -646,8 +647,10 @@ namespace Akka.Actor
     }
     public struct Envelope
     {
-        public object Message { get; set; }
-        public Akka.Actor.IActorRef Sender { get; set; }
+        public Envelope(object message, Akka.Actor.IActorRef sender, Akka.Actor.ActorSystem system) { }
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: I like this change a lot. Making the setters private and using a constructor instead.

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt (hunk 4 above)
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: Ah, I see now down below why that is....

Comment (alexvaluyskiy) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: I did the same what Akka JVM has

File: src/core/Akka/Actor/ActorCell.cs [hunk 6]
```diff
@@ -296,7 +296,7 @@ public void UseThreadContext(Action action)
             }
         }
 
-        public virtual void SendMessage(IActorRef sender, object message)
+        public virtual void SendMessage(Envelope message)
         {
             if (Mailbox == null)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/ActorCell.cs]: What is that? Akka JVM doesn't have this check

File: src/core/Akka/Actor/ActorCell.cs [hunk 7]
```diff
@@ -305,36 +305,41 @@ public virtual void SendMessage(IActorRef sender, object message)
...
 
-            var m = new Envelope
+                Dispatcher.Dispatch(this, message);
+            }
+            catch (Exception e)
             {
-                Sender = sender,
-                Message = message,
-            };
+                _systemImpl.EventStream.Publish(new Error(e, _self.Parent.ToString(), ActorType, "Swallowing exception during message send"));
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/ActorCell.cs]: Added missed exception handler

File: src/core/Akka/Util/TypeExtensions.cs [hunk 8]
```diff
@@ -37,6 +38,16 @@ public static bool Implements(this Type type, Type moreGeneralType)
         {
             return moreGeneralType.IsAssignableFrom(type);
         }
+
+        /// <summary>
+        /// Utility to be used by implementors to create a manifest from the type.
+        /// The manifest is used to look up the type on deserialization.
+        /// </summary>
+        /// <returns>Returns the type qualified name including namespace and assembly, but not assembly version.</returns>
+        public static string TypeQualifiedName(this Type type)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Util/TypeExtensions.cs]: We have the same method in Persistence. But we need it everywhere, not only in serializers

File: src/core/Akka/Actor/ActorCell.cs [hunk 9]
```diff
@@ -305,36 +305,43 @@ public virtual void SendMessage(IActorRef sender, object message)
...
-                        message = _systemImpl.Serialization.Deserialize(serialized, serializer.Identifier, message.GetType().AssemblyQualifiedName);
+                        Serializer serializer = _systemImpl.Serialization.FindSerializerFor(unwrapped);
+                        byte[] serialized = serializer.ToBinary(unwrapped);
+
+                        var manifestSerializer = serializer as SerializerWithStringManifest;
+                        if (manifestSerializer != null)
+                        {
+                            var manifest = manifestSerializer.Manifest(unwrapped);
+                            var deserialized = _systemImpl.Serialization.Deserialize(serialized, serializer.Identifier, manifest);
+                            message = new Envelope(deserialized, message.Sender);
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/ActorCell.cs]: Some tests don't work without this. But scala code doesn't have this assignment

**Pull request merged**